src/osd/osd_types.h

```diff
@@ -2664,6 +2667,26 @@ struct pg_log_t {
     return head.version - tail.version;
   }
 
+  list<pg_log_entry_t>::const_iterator find_entry(eversion_t v) const {
+    int fromhead = head.version - v.version;
+    int fromtail = v.version - tail.version;
+    list<pg_log_entry_t>::const_iterator p;
+    if (fromhead < fromtail) {
+      p = log.end();
+      --p;
+      while (p->version > v)
+	--p;
+    } else {
+      p = log.begin();
+      while (p->version < v)
+	++p;
+    }
```

Comment:
I don't like this optimization.  It only happens during startup, and it seems likely that read_log could be more clever if it really matters.  Just search the log from begin() unconditionally.  Also, check for end(), memory safety shouldn't rely on the parameter.